node/primitives/src/lib.rs

```diff
@@ -347,29 +347,39 @@ impl CollationResult {
 	}
 }
 
-/// Collation function.
-///
-/// Will be called with the hash of the relay chain block the parachain block should be build on and the
-/// [`ValidationData`] that provides information about the state of the parachain on the relay chain.
-///
-/// Returns an optional [`CollationResult`].
+/// A wrapper over a parachain collator.
 #[cfg(not(target_os = "unknown"))]
-pub type CollatorFn = Box<
-	dyn Fn(
-			Hash,
-			&PersistedValidationData,
-		) -> Pin<Box<dyn Future<Output = Option<CollationResult>> + Send>>
-		+ Send
-		+ Sync,
->;
+#[async_trait]
+pub trait Collator: Send + Sync {
+	/// Collation function.
+	///
+	/// Will be called with the hash of the relay chain block the parachain block should be build on and the
+	/// [`ValidationData`] that provides information about the state of the parachain on the relay chain.
+	///
+	/// Returns an optional [`CollationResult`].
+	async fn produce_collation(
+		&self,
+		relay_parent: Hash,
+		validation_data: &PersistedValidationData,
+	) -> Option<CollationResult>;
+
+	/// If a parachain consensus allows it (e.g. AuRa), figure out whether a collator is going to produce a
+	/// candidate on the child of the given relay parent. This allows a node to issue pre-connect requests
+	/// to a validators group.
+	///
+	/// The default implementation works for the rest chains such as PoW and returns `false`.
```

Comment:
I would prefer to avoid a default implementation since this can have quite serious performance implications, and favor being explicit.

From an API stability point of view, I think a default implementation is OK. but we should make it explicit at some point.

Reply:
> From an API stability point of view, I think a default implementation is OK. but we should make it explicit at some point.

Wait, given that this entire trait is a breaking change, we probably ought to just remove the default implementation.

But first I'd be curious to hear your thought process behind the default implementation in the first place.

Reply:
My reasoning is that parachains with consensus such as PoW cannot predict who is going to collate, so it could be left untouched. Then the behavior of collator protocol doesn't really change.

But probably we will implement the whole trait in cumulus anyway, so I agree here

Reply:
@rphmeier, for Moonbeam we use nimbus and can_author which relies on 2 inputs: `parachain parent hash` and `relay block number`.
I believe we could retrieve the block number from the block hash that is provided the the function. Otherwise we can change nimbus to rely on the relay hash.

How often (and when) will this api be called ?

Reply:
@crystalin It will be called on every relay parent for which parachain core is occupied (i.e. collator is not producing a candidate on this relay parent, but maybe? it will on child)

Reply:
Also, I will change the interface so that the function also takes persisted validation data, which will give you the relay parent block number.

Reply:
@slumber thank you, that will make it easier to integrate